### PR TITLE
identity provider directory implementations

### PR DESCRIPTION
**Pull request by calebdoxsey**
## Summary
This adds all of our existing identity provider directory implementations as external data sources.

## Related issues
Fixes https://github.com/pomerium/internal/issues/985

## Checklist
- [x] reference any related issues
- [ ] updated docs
- [x] updated unit tests
- [ ] updated UPGRADING.md
- [x] add appropriate tag (`improvement` / `bug` / etc)
- [x] ready for review

### CODE REVIEW

File: pkg/directory/http.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	groups, users, err := h.provider.GetDirectory(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get directory data: %w", err)
+	}
+	w.Header().Set("Content-Type", "application/zip")
+	return encodeBundle(w, groups, users)
+}
+
+func decodeBundle(r io.Reader) (groups []Group, users []User, err error) {
+	bs, err := io.ReadAll(r)
```
Comment (desimone) [inline pkg/directory/http.go:48]: Should we throw a `io.LimitReader` reader on this?

**Review by desimone — APPROVED** [review]
LGTM with a few suggestsions and comments. I largely reviewed the server / client / supporting code not the provider code directly since I assume that remains largely unchanged. LMK if you need a deeper review.

File: internal/bamboohr/ooo_test.go [hunk 2]
```diff
@@ -19,7 +21,10 @@ func TestOOO(t *testing.T) {
 			{Start: now.Add(-time.Second), End: now.Add(time.Second)},
 		}},
 	} {
+		tc := tc
 		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
```
[the remark this answers is not in the document]
Comment (calebdoxsey) [reply internal/bamboohr/ooo_test.go:26]: Yes the sub test can also run in parallel.

File: internal/server/server.go [hunk 3]
```diff
@@ -14,7 +15,8 @@ func RunHTTPServer(ctx context.Context, addr string, handler http.Handler) error
 		BaseContext: func(l net.Listener) context.Context {
 			return ctx
 		},
-		Handler: handler,
+		Handler:           handler,
+		ReadHeaderTimeout: time.Minute,
```
[the remark this answers is not in the document]
Comment (calebdoxsey) [reply internal/server/server.go:19]: linter flagged it

File: pkg/directory/http.go (hunk 1 above)
Comment (calebdoxsey) [reply pkg/directory/http.go:48]: this is only used for tests

**Comment by wasaga** [comment]
why `.pre-commit-config.yaml` was removed ?

File: internal/util/mapstructure.go [hunk 4]
```diff
@@ -92,11 +92,11 @@ func JSONNumberDecodeHook(srcT, dstT reflect.Type, data interface{}) (interface{
...
 		return val, nil
 	case float64:
-		if x := math.Trunc(val); x != val {
+		x := math.Trunc(val)
+		if x != val {
 			return nil, fmt.Errorf("JSON number must be represented as a whole number, got %f", val)
-		} else {
-			return fmt.Sprint(int(x)), nil
 		}
+		return fmt.Sprint(int(x)), nil
```
Comment (wasaga) [inline internal/util/mapstructure.go:99]: is that a manual change or code formatter?

File: pkg/directory/http.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to open zip file for reading: %w", err)
+	}
+
+	for _, file := range []struct {
+		name string
+		ptr  any
+	}{
+		{"pomerium.io/DirectoryGroup.json", &groups},
+		{"pomerium.io/DirectoryUser.json", &users},
```
Comment (wasaga) [inline pkg/directory/http.go:63]: would be great to have that defined as public constants so they could be referred from the console

File: scripts/download-azure/main.go [hunk 6]
```diff
@@ -56,7 +56,12 @@ func getWorkingDirectory() (string, error) {
 }
 
 func getDownloadURL(ctx context.Context) (string, error) {
```
Comment (wasaga) [inline scripts/download-azure/main.go:58]: maybe worth renaming this dir to something like `download-azure-ip-range`, as the repo grew in size and contains many disconnected things

**Review by wasaga — APPROVED** [review]

File: internal/util/mapstructure.go (hunk 4 above)
Comment (calebdoxsey) [reply internal/util/mapstructure.go:99]: The linter doesn't like `else` blocks.

**Comment by calebdoxsey** [comment]
The `pre-commit` referenced things in the Makefile that weren't there.

**Pull request merged**